### PR TITLE
docs: fix type warnings

### PR DESCRIPTION
**Pull request by MariaSolOs**
"Types are good." - @sid-6581

### CODE REVIEW

File: runtime/doc/treesitter.txt [hunk 1]
```diff
@@ -682,8 +682,7 @@ inspect_tree({opts})                           *vim.treesitter.inspect_tree()*
     Can also be shown with `:InspectTree`.                      *:InspectTree*
 
     Parameters: ~
-      • {opts}  (table|nil) Optional options table with the following possible
-                keys:
+      • {opts}  InspectTreeOpts ? Optional options table with the following possible keys:
```
Comment (MariaSolOs) [inline runtime/doc/treesitter.txt:685]: I'm not sure why this type wasn't public in the first place. If intended I can undo the change.

File: runtime/lua/vim/keymap.lua [hunk 2]
```diff
@@ -79,6 +79,9 @@ function keymap.set(mode, lhs, rhs, opts)
   end
 end
 
+---@class KeymapDelOpts
```
Comment (lewis6991) [inline runtime/lua/vim/keymap.lua:82]: Make sure to namespace any classes you add.

File: runtime/lua/vim/uri.lua [hunk 3]
```diff
@@ -82,7 +82,7 @@ function M.uri_from_bufnr(bufnr)
   local fname = vim.api.nvim_buf_get_name(bufnr)
   local volume_path = fname:match('^([a-zA-Z]:).*')
   local is_windows = volume_path ~= nil
-  local scheme
+  local scheme ---@type string
```
Comment (lewis6991) [inline runtime/lua/vim/uri.lua:85]: `string?`

File: runtime/lua/vim/keymap.lua (hunk 2 above)
Comment (MariaSolOs) [reply runtime/lua/vim/keymap.lua:82]: I had actually done that [initially](https://github.com/neovim/neovim/compare/b7453bf08af2634a43ce972c353ea61e97f80ef5..49cb777a91978a78656d5afaa7315bc84b163653), but the `doxygen` script is messed up and adds extra spaces that make CI fail :/

Comment (lewis6991) [reply runtime/lua/vim/keymap.lua:82]: Then we need to fix that or hold off on adding types that are exposed.

File: runtime/doc/treesitter.txt (hunk 1 above)
Comment (MariaSolOs) [reply runtime/doc/treesitter.txt]: Undoing the change since `doxygen` is broken and I cannot it namespace it.

File: runtime/lua/vim/keymap.lua [hunk 4]
```diff
@@ -44,8 +44,8 @@ function keymap.set(mode, lhs, rhs, opts)
     opts = { opts, 't', true },
   })
 
-  opts = vim.deepcopy(opts) or {}
-  mode = type(mode) == 'string' and { mode } or mode
+  opts = vim.deepcopy(opts or {})
+  mode = type(mode) == 'string' and { mode } or mode ---@cast mode string[]
```
Comment (lewis6991) [inline runtime/lua/vim/keymap.lua:48]: `@cast` usually goes on its own line. You might want to use `@as` here.

File: runtime/lua/vim/keymap.lua [hunk 5]
```diff
@@ -67,9 +67,9 @@ function keymap.set(mode, lhs, rhs, opts)
 
   if opts.buffer then
     local bufnr = opts.buffer == true and 0 or opts.buffer
-    opts.buffer = nil
+    opts.buffer = nil ---@type integer?
     for _, m in ipairs(mode) do
-      vim.api.nvim_buf_set_keymap(bufnr, m, lhs, rhs, opts)
+      vim.api.nvim_buf_set_keymap(bufnr --[[@as integer]], m, lhs, rhs, opts)
```
Comment (lewis6991) [inline runtime/lua/vim/keymap.lua:72]: Shouldn't `bufnr` be typed as `integer`?

File: runtime/lua/vim/keymap.lua (hunk 2 above)
Comment (lewis6991) [reply runtime/lua/vim/keymap.lua:82]: Btw one trick I've used to exclude types from the docs is to use `--- @cast opts Myopts` at the top of the function body.

File: runtime/lua/vim/treesitter.lua [hunk 6]
```diff
@@ -127,7 +127,7 @@ function M.get_parser(bufnr, lang, opts)
       )
     end
   elseif parsers[bufnr] == nil or parsers[bufnr]:lang() ~= lang then
-    parsers[bufnr] = M._create_parser(bufnr, lang, opts)
+    parsers[bufnr] = M._create_parser(bufnr, lang --[[@as string]], opts)
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter.lua:130]: This shouldn't be needed. Lang must be losing its type further up.

Comment (MariaSolOs) [reply runtime/lua/vim/treesitter.lua:130]: It's not that it's losing it's type, but the parameter type is `string|nil` but the checks above should ensure that in here we have a valid `lang`.

Comment (lewis6991) [reply runtime/lua/vim/treesitter.lua:130]: If you need to remove `nil` from a type then best to use `assert(var)`.

**Review by lewis6991 — APPROVED** [review]

File: runtime/lua/vim/keymap.lua [hunk 7]
```diff
@@ -89,7 +94,7 @@ end
 --- ```
 ---
 ---@param opts table|nil A table of optional arguments:
----                      - "buffer": (number|boolean) Remove a mapping from the given buffer.
+---                      - "buffer": (integer|boolean) Remove a mapping from the given buffer.
```
Comment (zeertzjq) [inline runtime/lua/vim/keymap.lua:97]: Why is this only changed for `vim.keymap.del()` but not `vim.keymap.set()`?

Comment (MariaSolOs) [reply runtime/lua/vim/keymap.lua:97]: No reason in particular. I was just fixing warnings and there were none for this method, but I'll make the change as well.

File: runtime/lua/vim/secure.lua [hunk 8]
```diff
@@ -108,6 +108,11 @@ function M.read(path)
   return contents
 end
 
+---@class TrustOpts
```
Comment (lewis6991) [inline runtime/lua/vim/secure.lua:111]:
```suggestion
---@class vim.trust.opts
```

**Review by lewis6991 — APPROVED** [review]

File: runtime/lua/vim/keymap.lua [hunk 9]
```diff
@@ -79,6 +81,9 @@ function keymap.set(mode, lhs, rhs, opts)
   end
 end
 
+---@class KeymapDelOpts
```
Comment (lewis6991) [inline runtime/lua/vim/keymap.lua:84]: Can't see where this is used. Also needs `vim.` prefix.

Comment (MariaSolOs) [reply runtime/lua/vim/keymap.lua:84]: Good catch. I don't think we need it anymore.

**Pull request merged**